### PR TITLE
Feature: Dedicated list configuration tools (create-list, update-list-configuration) #1195

### LINKED ISSUE #1195: feature: list configuration tools (for Attio lists)
## Priority
P2

## Area
area:api:lists

## Summary
Expose scoped MCP tools for creating and updating Attio list configurations.

## Problem to solve
Agents can currently work with existing Attio lists through list-entry tools (`get-lists`, `get-list-details`, `get-list-entries`, `add-record-to-list`, `update-list-entry`, `remove-record-from-list`, and filters), but there is no obvious first-class MCP tool for creating a new list or modifying list configuration/fields.

This blocks operational workflows where the user asks an agent to set up a new Attio list for a sales or recruiting process. The token may have `list_configuration:read-write`, but without a scoped MCP surface the safer behavior is to refuse or ask the user to create the list manually in the Attio UI.

## Proposed solution
Add explicit list-configuration tools with a narrow, typed input surface, for example:

- `create-list` / `create_list`
- `update-list-configuration` / `update_list_configuration`
- optional `create-list-attribute` / `update-list-attribute` if Attio models list fields separately from list metadata

The tools should support preview/dry-run friendly output where possible, validate allowed parent object types and field definitions before writing, and return clear errors for unsupported schema/access-control combinations.

This should be separate from destructive object-configuration tools. Creating a list for an existing object is materially safer than deleting objects or attributes, but it still needs guardrails because bad list schemas create workspace clutter and confusing operator UX.

## Alternatives considered
Keep list configuration as a manual UI-only step while MCP only manages entries on existing lists. This is safe but blocks otherwise straightforward agent workflows and forces humans to copy list structures by hand.

Use raw Attio REST calls from agents when the token has `list_configuration:read-write`. This is weaker because it bypasses MCP validation, consistent errors, audit-friendly tool names, and any future compatibility handling for list access controls.

## Impact
Affected workflows: agents setting up sales, recruiting, upgrade, support, or campaign tracking lists in Attio.

Severity: blocks workflow setup, not day-to-day entry management once a list exists.

Frequency: recurring whenever a new operational list/process is introduced.

Consequence: manual UI setup, inconsistent list structures, and agent refusals even when the underlying token has the necessary Attio scope.

## Acceptance criteria
- [ ] MCP exposes a documented tool to create an Attio list for a supported parent object type.
- [ ] MCP exposes a documented tool or workflow to update list configuration/fields, or explicitly documents why specific configuration changes are unsupported.
- [ ] Tools validate parent object type, list name, field definitions, and access-control inputs before sending writes to Attio.
- [ ] User-facing errors distinguish unsupported input, insufficient token scope, Attio permission/access-control failure, and generic API failure.
- [ ] Tests cover successful create/list-config update paths using mocked Attio API responses.
- [ ] Docs list the new tool names and a minimal example payload.

## Evidence/examples
Current `attio-b2b` MCP tool list includes entry-level list tools but no obvious list-configuration create/update tool:

```text
get-lists
get-list-details
get-list-entries
add-record-to-list
update-list-entry
remove-record-from-list
filter-list-entries
advanced-filter-list-entries
get-record-list-memberships
```

Adjacent existing issue: #1148 tracks support for current Attio list access-control fields and restricted-workspace errors. This request is specifically about exposing first-class list create/update configuration tools.

Attio docs:
- https://docs.attio.com/rest-api/endpoint-reference/lists/create-a-list
- https://attio.com/help/reference/managing-your-data/lists/lists-access

## Additional information
This came from a real agent workflow where the safe local policy had to refuse list creation even though the service token has `list_configuration:read-write`. A scoped MCP tool would let downstream agents relax that policy without allowing raw schema/configuration writes.

### PR DESCRIPTION
**Pull request by kesslerio**
## Summary

Implements dedicated list configuration tools with smart defaults, template expansion, immutable field detection, and dry-run mode.

### U1: Shared List Configuration Validator
- **`src/services/lists/types.ts`** — `NormalizedListResponse`, `ListTemplate`, `CategorizedListError`, `IMMUTABLE_LIST_FIELDS`, `normalizeListResponse()`
- **`src/services/lists/list-templates.ts`** — Static template catalog (`sales_pipeline`, `recruiting_tracker`, `support_queue`) with `expandTemplate()` merging caller overrides onto defaults
- **`src/services/lists/ListConfigurationValidator.ts`** — Shared validator with:
  - `validateParentObject()` — auto-resolves valid objects via `GET /objects` with 60s in-memory cache
  - `detectImmutableFields()` — rejects `parent_object` updates before API call
  - `expandTemplate()` — delegates to list-templates
  - `normalizeResponse()` — consistent agent-friendly shape
  - `categorizeError()` — maps API errors to actionable categories with next-step guidance

### U2: Dedicated `create-list` Tool
- Tool config + definition in `src/handlers/tool-configs/lists.ts`
- Dispatcher handler `handleCreateListOperation` in `src/handlers/tools/dispatcher/operations/lists.ts`
- Template expansion before validation (R8), parent-object auto-resolution, dry-run preview, normalized response

### U3: Dedicated `update-list-configuration` Tool
- Tool config + definition in `src/handlers/tool-configs/lists.ts`
- Dispatcher handler `handleUpdateListConfigurationOperation` in `src/handlers/tools/dispatcher/operations/lists.ts`
- Immutable field detection (rejects `parent_object`), dry-run preview, categorized errors

### U4: Universal Path Hardening
- `ListCreateStrategy` now validates `parent_object` via shared validator before API call
- `ListUpdateStrategy` now detects immutable fields via shared validator before API call
- Consistent error handling between dedicated and universal paths

## Test Coverage
- 50 unit tests across 4 test files:
  - `list-templates.test.ts` (11 tests)
  - `ListConfigurationValidator.test.ts` (23 tests)
  - `list-configuration-tools.test.ts` (9 tests)
  - `list-strategies-hardening.test.ts` (7 tests)
- Full offline suite: 3522 tests passing

## Quality
- Typecheck: clean
- Lint: clean (0 errors, 1 pre-existing warning in tool-types.ts)

Closes #1195

### CODE REVIEW

**Review by kesslerio — COMMENTED** [review]
# PR #1196 Review: Dedicated List Configuration Tools

## Summary Assessment

The PR delivers solid value: dedicated `create-list` and `update-list-configuration` tools with template expansion, immutable field detection, parent-object validation, and dry-run mode. The universal path hardening (U4) is a good defensive measure. However, there are several issues ranging from P1 (overengineering, missing test coverage for the actual dispatcher handlers) to P2 (fragile error categorization, dynamic import anti-pattern) that should be addressed before merge.

---

## File-by-File Findings

### `src/handlers/tools/dispatcher/operations/lists.ts` (lines 1180–1425)

**P1 — Dynamic imports used where static imports are safe**

Both `handleCreateListOperation` and `handleUpdateListConfigurationOperation` use `await import(...)` for `ListConfigurationValidator`, `createList`/`updateList`, and `UniversalValidationError`. The comment says "avoid circular dependency at module load" but:

- `ListConfigurationValidator` is in `src/services/lists/` — there is no circular dependency path back to the dispatcher.
- `createList`/`updateList` are in `src/objects/lists/base.js` — also no circular path.
- `UniversalValidationError` is already statically imported by `ListConfigurationValidator.ts` itself.

Dynamic imports add cold-start latency on first invocation and make the code harder to trace/debug. They should be converted to static imports at the top of the file, consistent with every other handler in this same file.

```typescript
// Replace all dynamic imports with static imports at the top:
import { ListConfigurationValidator } from '@/services/lists/ListConfigurationValidator.js';
import { createList, updateList } from '@/objects/lists/base.js';
import { UniversalValidationError } from '@/handlers/tool-configs/universal/errors/validation-errors.js';
```

**P1 — No test coverage for the actual dispatcher handlers**

`test/services/lists/list-configuration-tools.test.ts` tests the `ListConfigurationValidator` methods directly but **never calls `handleCreateListOperation` or `handleUpdateListConfigurationOperation`**. These are the actual code paths users hit. The test file name implies it tests the tools, but it only tests the validator in isolation. This leaves the parameter extraction, dry-run branching, error categorization in the catch block, and `formatResult` integration completely untested.

Add tests that construct a `CallToolRequest`-like object and call the actual handler functions, mocking `ListConfigurationValidator` and `createList`/`updateList`.

**P2 — Duplicated error-handling boilerplate in catch blocks**

The catch blocks in `handleCreateListOperation` (lines 1279–1314) and `handleUpdateListConfigurationOperation` (lines 1389–1423) are nearly identical — dynamic import of `UniversalValidationError`, dynamic import of `ListConfigurationValidator`, categorize, build message, build response data. Extract a helper:

```typescript
async function handleListToolError(error: unknown, path: string, method: string) {
  const isUserError = error instanceof UniversalValidationError;
  const status = isUserError ? error.httpStatusCode : undefined;
  const categorized = ListConfigurationValidator.categorizeError(error);
  const errorMessage = categorized
    ? `${categorized.message} (Next step: ${categorized.suggested_next_step})`
    : error instanceof Error ? error.message : 'Unknown error';
  const responseData = hasResponseData(error) ? error.response.data : {};
  if (status) (responseData as Record<string, unknown>).status = status;
  return createErrorResult(new Error(errorMessage), path, method, responseData);
}
```

**P2 — Dry-run preview fabricates an `AttioList` with incorrect shape**

Lines 1253–1262 construct a fake `AttioList` via `as` cast. The `title` field is set to `listAttributes.name`, but `object_slug` is set to `listAttributes.parent_object`. If template expansion changed `parent_object`, the preview shows the expanded value — which is correct. But `workspace_id: 'preview'` and `created_at: new Date().toISOString()` are fabricated. The `...listAttributes` spread then overwrites `object_slug` with `parent_object` if that key exists in `listAttributes`, creating a confusing shape. Since dry-run is a preview, this is acceptable but the spread ordering should be explicit — put `...listAttributes` first, then override the known metadata fields:

```typescript
const preview = ListConfigurationValidator.normalizeResponse(
  {
    ...listAttributes,
    id: { list_id: 'dry-run-preview' },
    title: listAttributes.name as string,
    object_slug: listAttributes.parent_object as string,
    workspace_id: 'preview',
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
  } as AttioList,
  true
);
```

---

### `src/services/lists/ListConfigurationValidator.ts`

**P2 — `extractStatus` doesn't handle `AttioApiError`**

The `extractStatus` helper (lines 255–265) only checks for axios-style `error.response.status` or a bare `status` property. The codebase has a rich hierarchy of `AttioApiError` subclasses (`AuthorizationError` for 403, `AuthenticationError` for 401, etc.) that carry `status` as a direct property. While the bare `status` check may catch some cases, the type assertion `(error as { status: number }).status` could pick up unrelated `status` properties. Prefer an explicit check:

```typescript
function extractStatus(error: unknown): number | undefined {
  if (!error) return undefined;
  if (error instanceof AttioApiError) return error.status;
  if (typeof error === 'object' && 'response' in error) {
    const resp = (error as { response?: { status?: number } }).response;
    return resp?.status;
  }
  return undefined;
}
```

This requires importing `AttioApiError` from `@/errors/api-errors.js`.

**P2 — `categorizeError` uses fragile string matching for `UNSUPPORTED_INPUT`**

Lines 229–234 check `message.includes('Invalid')` — this will match any error message containing the word "Invalid", including legitimate API failures that happen to use that word. The `message.includes('is required')` and `message.includes('must be')` patterns are similarly broad. Consider narrowing to known Attio API error patterns or checking HTTP status 400 first.

**P3 — Module-level mutable cache variable**

`cachedObjects` (line 34) is a module-level `let` that makes testing harder and could cause state leakage between requests in long-running processes. The existing `invalidateObjectCache()` export addresses testing, but consider encapsulating the cache within the class or using a `WeakRef`-based approach for production safety.

---

### `src/services/lists/types.ts`

**P2 — `normalizeListResponse` uses `raw.name || raw.title` but `title` is required on `AttioList`**

The `AttioList` type declares `title: string` (required) and `name?: string` (optional). The normalization at line 63 does `raw.name || raw.title || ''`. Since `title` is always present, the fallback chain is backwards — it should prefer `title` (guaranteed) over `name` (optional):

```typescript
const name = raw.title || raw.name || '';
```

This also aligns with `ListUpdateStrategy` line 34 which does `name: list.name || list.title`.

---

### `src/services/lists/list-templates.ts`

**P3 — `expandTemplate` spread ordering silently drops template `description`**

Lines 84–94: The `...overrides` spread happens before the `if (!overrides.description && template.description)` check. But if `overrides.description` is explicitly set to `''` (empty string), the `!overrides.description` check is truthy, so the template description overwrites the caller's explicit empty string. This is a minor semantic issue — if a caller explicitly sets `description: ''`, they probably want it empty. Fix:

```typescript
if (!('description' in overrides) && template.description) {
  expanded.description = template.description;
}
```

---

### `src/handlers/tool-configs/lists.ts`

**P3 — `create-list` `inputSchema` has `template` enum hardcoded**

Line 1052: The `template` property enum is `['sales_pipeline', 'recruiting_tracker', 'support_queue']`. When new templates are added to `LIST_TEMPLATES`, this enum must be manually updated in sync. Consider deriving the enum from `getTemplateNames()` or at minimum adding a comment noting the coupling.

---

### `src/services/create/strategies/ListCreateStrategy.ts`

**P2 — Inconsistent `UniversalValidationError` import path**

Line 11 imports from `@/handlers/tool-configs/universal/schemas.js` (a re-export barrel), while `ListConfigurationValidator.ts` imports from `@/handlers/tool-configs/universal/errors/validation-errors.js` (the source). Per AGENTS.md: "Path aliases over relative imports" — both use `@/` which is fine, but the import paths should be consistent. Prefer the direct source path `errors/validation-errors.js` to avoid an extra indirection hop, matching the validator's pattern.

---

### `src/services/update/strategies/ListUpdateStrategy.ts`

**P2 — Same inconsistent import path as ListCreateStrategy**

Line 5 imports `UniversalValidationError` from `schemas.js` instead of `errors/validation-errors.js`. Same fix as above.

---

### Test Files

**P1 — `list-configuration-tools.test.ts` doesn't test the actual handler functions**

As noted above, this file tests `ListConfigurationValidator` methods in isolation but never invokes `handleCreateListOperation` or `handleUpdateListConfigurationOperation`. The file name is misleading. Either rename to reflect what it tests, or add proper handler-level tests.

**P2 — No test for `extractStatus` function**

The `extractStatus` helper in `ListConfigurationValidator.ts` has no direct test coverage. The `categorizeError` tests in `ListConfigurationValidator.test.ts` happen to exercise it indirectly via objects with `response.status`, but there's no test for the `AttioApiError` path or the bare `status` property path.

---

## Overengineering Assessment

The PR is **moderately overengineered** in two areas:

1. **`ListConfigurationValidator` as a static class with 5 methods** — Three of the five methods (`expandTemplate`, `normalizeResponse`, `categorizeError`) are pure delegation/wrappers that add no logic beyond calling the underlying function. The class exists primarily as a namespace, which TypeScript modules already provide. The static class pattern adds indirection without benefit. Consider exporting the functions directly from their modules and having the dispatcher call them directly.

2. **Error categorization with `CategorizedListError` + `ListErrorCategory` enum** — This is a 4-category error taxonomy with suggested next steps, but only two callers (the two catch blocks) consume it, and the categorization logic is fragile string matching. A simpler approach: just throw `UniversalValidationError` with a `suggestion` field (which already exists) and let the existing error-handling pipeline handle it. The categorization layer adds complexity without proportional value at this stage.

---

## Severity Summary

| Severity | Count | Key Issues |
|----------|-------|------------|
| P0 | 0 | — |
| P1 | 2 | Dynamic imports where static are safe; no handler-level test coverage |
| P2 | 7 | Duplicated catch blocks; `extractStatus` misses `AttioApiError`; fragile string matching in `categorizeError`; `normalizeListResponse` title/name precedence; inconsistent import paths (×2); missing `extractStatus` test |
| P3 | 3 | Module-level mutable cache; template enum coupling; `expandTemplate` empty-string description edge case |

**Recommendation**: Request changes for the P1s (static imports, handler tests) and the `normalizeListResponse` precedence fix. The P2s are strongly encouraged but not blocking.

**Comment by kesslerio** [comment]
## Response to Review #4286083568

All P1 and P2 issues addressed in commit `bbd53d39`:

### P1 Fixes
- **Dynamic imports → static imports**: Converted all `await import(...)` in `handleCreateListOperation` and `handleUpdateListConfigurationOperation` to static top-level imports. Verified no circular dependency exists.
- **No handler-level test coverage**: Added `test/services/lists/list-config-handler.test.ts` with 11 tests that call the actual handler functions directly.

### P2 Fixes
- **Duplicated catch blocks**: Extracted `handleListToolError()` helper.
- **Dry-run spread ordering**: Fixed to `...listAttributes` first, then override metadata fields.
- **`extractStatus` misses `AttioApiError`**: Now checks `instanceof AttioApiError` → `statusCode` before axios fallback.
- **Fragile string matching in `categorizeError`**: Added `httpStatus === 400` check first; removed broad `message.includes("Invalid")`.
- **`normalizeListResponse` title/name precedence**: Fixed to `raw.title || raw.name || ""`.
- **Inconsistent import paths**: Both strategies now import from `errors/validation-errors.js` directly.
- **Missing `extractStatus` test**: Added `AttioApiError` categorization test.
- **`normalizeListResponse` defensive**: Added null/undefined guard.

### P3 (deferred, not blocking)
- Module-level mutable cache, template enum coupling, expandTemplate empty-string edge case.

@codex for re-review.

File: src/services/lists/ListConfigurationValidator.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+      };
+    }
+
+    // Check for unsupported input patterns — prefer HTTP status 400 over string matching
+    const httpStatus = extractStatus(error);
+    if (
+      httpStatus === 400 ||
+      message.includes('Cannot find attribute') ||
+      message.includes('is required') ||
+      message.includes('must be')
```
[the remark this answers is not in the document]
Comment (kesslerio) [reply src/services/lists/ListConfigurationValidator.ts]: Fixed in commit `324e6601`. Added `message.startsWith('Invalid list attributes')` to catch the wrapped 400 errors from `base.ts` that lose `response.status`. This is narrower than the previous `message.includes('Invalid')` — only matches the specific error prefix from the list CRUD layer.

**Comment by kesslerio** [comment]
@codex re-review — P2 wrapped-400 issue addressed in `324e6601`.
